op-node/rollup/derive/l1_block_info.go

```diff
@@ -133,29 +143,141 @@ func (info *L1BlockInfo) UnmarshalBinary(data []byte) error {
+// +---------+--------------------------+
+
+func (info *L1BlockInfo) marshalBinaryEcotone() ([]byte, error) {
+	w := bytes.NewBuffer(make([]byte, 0, L1InfoEcotoneLen))
+	if err := solabi.WriteSignature(w, L1InfoFuncEcotoneBytes4); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(w, binary.BigEndian, info.BasefeeScalar); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(w, binary.BigEndian, info.BlobBasefeeScalar); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(w, binary.BigEndian, info.SequenceNumber); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(w, binary.BigEndian, info.Time); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(w, binary.BigEndian, info.Number); err != nil {
+		return nil, err
+	}
+	if err := solabi.WriteUint256(w, info.Basefee); err != nil {
+		return nil, err
+	}
+	if err := solabi.WriteUint256(w, info.BlobBasefee); err != nil {
+		return nil, err
+	}
+	if err := solabi.WriteHash(w, info.BlockHash); err != nil {
+		return nil, err
+	}
+	if err := solabi.WriteAddress(w, info.BatcherAddr); err != nil {
+		return nil, err
+	}
+	return w.Bytes(), nil
+}
+
+func (info *L1BlockInfo) unmarshalBinaryEcotone(data []byte) error {
+	if len(data) != L1InfoEcotoneLen {
+		return fmt.Errorf("data is unexpected length: %d", len(data))
+	}
+	r := bytes.NewReader(data)
+
+	var err error
+	if _, err := solabi.ReadAndValidateSignature(r, L1InfoFuncEcotoneBytes4); err != nil {
+		return err
+	}
+	if err := binary.Read(r, binary.BigEndian, &info.BasefeeScalar); err != nil {
+		return fmt.Errorf("invalid ecotone l1 block info format")
+	}
+	if err := binary.Read(r, binary.BigEndian, &info.BlobBasefeeScalar); err != nil {
+		return fmt.Errorf("invalid ecotone l1 block info format")
+	}
+	if err := binary.Read(r, binary.BigEndian, &info.SequenceNumber); err != nil {
+		return fmt.Errorf("invalid ecotone l1 block info format")
+	}
+	if err := binary.Read(r, binary.BigEndian, &info.Time); err != nil {
+		return fmt.Errorf("invalid ecotone l1 block info format")
+	}
+	if err := binary.Read(r, binary.BigEndian, &info.Number); err != nil {
+		return fmt.Errorf("invalid ecotone l1 block info format")
+	}
+	if info.Basefee, err = solabi.ReadUint256(r); err != nil {
+		return err
+	}
+	if info.BlobBasefee, err = solabi.ReadUint256(r); err != nil {
+		return err
+	}
+	if info.BlockHash, err = solabi.ReadHash(r); err != nil {
+		return err
+	}
+	if info.BatcherAddr, err = solabi.ReadAddress(r); err != nil {
+		return err
+	}
+	if !solabi.EmptyReader(r) {
+		return errors.New("too many bytes")
+	}
+	return nil
+}
+
+// isEcotoneButNotFirstBlock returns whether the specified block is subject to the Ecotone upgrade,
+// but is not the very first such block.
+func isEcotoneButNotFirstBlock(rollupCfg *rollup.Config, l2BlockTime uint64) bool {
+	return rollupCfg.IsEcotone(l2BlockTime) &&
+		(l2BlockTime < rollupCfg.BlockTime || rollupCfg.IsEcotone(l2BlockTime-rollupCfg.BlockTime))
```

Comment:
Maybe we can add `IsEcotoneActivationBlock` (or some variation to rollup.Cfg)? It feels like we use it once or twice.

Reply:
added.